app/lib/search/heap.dart

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A data structure that keep satisfies the "heap property". This property
+/// dictates that in a max-heap, each node's value is greater than or equal
+/// to its children's values, and in a min-heap, each node's value is less
+/// than or equal to its children's values.
+///
+/// The provided comparator decides which kind of heap is being built.
+class _Heap<T> {
+  final Comparator<T> _compare;
+  final _items = <T>[];
+
+  _Heap(this._compare);
+
+  int get length => _items.length;
+
+  void _pushDown(int index) {
+    final maxLength = _items.length;
+    final item = _items[index];
+    while (index < maxLength) {
+      final leftIndex = (index << 1) + 1;
+      if (leftIndex >= maxLength) {
+        return;
+      }
+      var childIndex = leftIndex;
+      final rightIndex = leftIndex + 1;
+      if (rightIndex < maxLength &&
+          _compare(_items[leftIndex], _items[rightIndex]) > 0) {
+        childIndex = rightIndex;
+      }
+      if (_compare(item, _items[childIndex]) <= 0) {
+        return;
+      }
+      _items[index] = _items[childIndex];
+      _items[childIndex] = item;
+      index = childIndex;
+    }
+  }
+
+  bool _isValidHeap() {
+    for (var i = 1; i < _items.length; i++) {
+      final parentIndex = (i - 1) >> 1;
+      if (_compare(_items[parentIndex], _items[i]) > 0) {
+        print(parentIndex);
+        print(_items);
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+/// Builds a sorted list of the top-k items using the provided comparator.
+///
+/// The algorithm uses min-heap to select the top-k items, and then builds
+/// a max-heap and uses heap sort to return the items in descending order.
```

Comment:
Is this description correct?

I only see a single heap in the code. Also, why would you use a min-heap for selecting top-k items?

Reply:
Ah, no, this is now obsolete, will update.